### PR TITLE
initial commit of SMARTS NN notebook

### PR DESCRIPTION
**Pull request by zclaytor**

### CODE REVIEW

**Comment by cmurray-astro** [comment]
Hi @zclaytor! Apologies for the delay in getting this review started. I've gone through the notebook and I have a few questions:

1. First, I ran into an error with using the "001/" group of simulated light curves (an index error with one of the files, shown in the comments of the notebook pushed here). When I edited to "002/" it all ran well. This is something we should verify with the HLSP ingest side.
2. It would be great to understand a bit more what your goals are for the notebook? If you can add a few words of comments before each cell, that will help me a lot with drafting the rest of the notebook. 
3. Do you expect the plot at the end of the notebook to look correlated? As with other Hello Universe examples, we don't promise that on-the-fly training with small datasets will get good results, but I just want to check with you that that's what you had in mind?

Thanks again for submitting this, looking forward to hearing from you!

**Comment by zclaytor** [comment]
Hi @cmurray-astro! Thanks for your questions. 

1. I isolated the issue to a corrupted file, simulation #408. I reran the sim and have a new tar file to upload. Should I go ahead and upload it?
2. I've added some cell comments following the "Classifying TESS Flares" notebook as an example. Hopefully this clarifies the goals but please let me know if I should add more.
3. It's totally fine that the plot at the end isn't correlated. Since we're just going over a handful of training examples, I don't expect the tutorial CNN to generalize well.

Thanks!

**Comment by cmurray-astro** [comment]
Hi Zach, thanks so much for the update! Yes, please go ahead and upload the updated tar file, and I'll process the HLSP update. In the meantime I'll review the notebook!

**Comment by zclaytor** [comment]
Great, thanks! I uploaded the new tar file. I have also verified that the rest of the files are free from this error.

**Comment by cmurray-astro** [comment]
hi @zclaytor, I'm working through the notebook, and am running into issues with the training step. Can you confirm which SMARTS datasets you downloaded+used for this case, and also please let me know which versions of the required python packages you are running (really just torch)

thanks!

**Comment by cmurray-astro** [comment]
and I should clarify -- the issue is that the notebook kernel dies when it hits the `output = model(data)` step in the train_epoch loop

**Comment by zclaytor** [comment]
After a very long delay, I have fixed the issues in the notebook that Claire identified before, and added some analysis at the end to explore the performance of the neural network. The notebook runs for me starting with the third cell, but the editorial PEP8 check cells at the top cause the notebook to fail the automated checks. I think they might be old or outdated?

**Comment by zclaytor** [comment]
I deleted the editorial cells to see if they were the problem. The validation check now fails on Cell 7 with the following output:

```
=================================== FAILURES ===================================
_ notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb::Cell 7 _
Notebook cell execution failed
Cell 7: Timeout of 5 seconds exceeded waiting for output.
Input:
print("Begin training:")
# Initialize CNN
model = ConvNet()
# Train CNN
weights, train_p_loss, val_p_loss = train(model, train_loader, val_loader,
    early_stopping_patience=10, num_epochs=100)
=========================== short test summary info ============================
FAILED notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb::Cell 7
============== 1 failed, 7 passed, 5 xfailed in 591.80s (0:09:51) ==============
Error: Process completed with exit code 1.
```

The cell begins producing output (from the print line) immediately when I run it in TIKE, so I'm not entirely sure what's failing here and why.

I've opened a STARS ticket with Notebook CI folks for help (RITM0166003).

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb [hunk 1]
```diff
@@ -0,0 +1,1019 @@
...
+    "plt.hist(frac, bins=20)\n",
+    "plt.xlabel(\"Fractional Period Uncertainty\");"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f68d6882-9de6-4e3f-a7b0-352a285eb94d",
+   "metadata": {},
+   "source": [
+    "By 20 epochs we start to see a dip forming at sigma/period ~ 0.5 (this becomes more pronounced the longer we train), so let's use that to filter the results, weeding out \"bad\" predictions."
```
Comment (eteq) [inline notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:919]: What makes you say "by 20 epochs"?  I don't see any plots or the like that demonstrate this.

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb [hunk 2]
```diff
@@ -0,0 +1,1019 @@
...
+   ],
+   "source": [
+    "print(\"Begin training:\")\n",
+    "\n",
+    "# Initialize CNN\n",
+    "model = ConvNet()\n",
+    "\n",
+    "# Train CNN\n",
+    "weights, train_p_loss, val_p_loss = train(model, train_loader, val_loader,\n",
+    "    early_stopping_patience=10, num_epochs=num_epochs)"
```
Comment (eteq) [inline notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:793]: Maybe here or in a follow-on cell you should plot the loss over epochs?  It will help the reader understand how things are evolving (especially given you say things below about it)

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb [hunk 3]
```diff
@@ -0,0 +1,1019 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "058fba43-4638-4967-a5f7-064d5488afaa",
+   "metadata": {},
+   "source": [
+    "With the \"bad\" predictions filtered out, the predictions look much better, even after only 20 epochs. Remember that the uncertainty is predicted from the quality of the data, so this kind of cut can be applied to predictions on real data as well. \n",
+    "\n",
+    "Finally, let's define an accuracy metric to measure the performance. For this example, we'll use the root-mean-squared (RMS) error."
```
Comment (eteq) [inline notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:955]: This is a bit confusing because this is basically the MSE loss function (technically I guess the sqrt of the MSE), but here you use the phrase "accuracy metric" - do you mean something different here?  If not, it might be a good time to emphasize that the loss function *is* an accuract metric.

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb [hunk 4]
```diff
@@ -0,0 +1,1019 @@
...
+    "      f\"Filtered prediction count:   {len(pred_filtered)}\\n\"\n",
+    "      f\"RMS error: {rms_error(true_filtered, pred_filtered):.2f} days.\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "508fe865-deac-4253-b18d-5221616e277f",
+   "metadata": {},
+   "source": [
+    "When the predictions are filtered, we retain only about 10-15% of the sample, but the accuracy improves significantly. Some takeaway notes:\n",
```
Comment (eteq) [inline notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:991]: I think you need a bit more of a summary.  Your "takeaway notes" here are basically "what someone who already understands the general method might learn from this specific example", but what you really want is a summary for someone who doesn't really know what a CNN is or why this method might be useful.

This might be as simple as just something like adding a note "CNNs combined with wavelets make reasonable predictions for TESS light curves if trained on data with different systematics."  Which I think is your main point?  But if not, please decide what it is and write it here!

**Review by eteq — CHANGES_REQUESTED** [review]
Sorry for the long delay on getting to this review.  I have a bunch of inline comments, but three bigger-picture items:

* You don't really say anywhere what TESS is, or why SMARTS data has anything to do with TESS.  I think I know the answer, but the average reader is not necessarily coming in with any knowledge of the mission.  So I think you need to explain what TESS is (where "explain" might just be "TESS makes light curves of stars in a particular band.  For more info see <link goes here>"), and similar for SMARTS.
* You've left in all the outputs of this notebook.  That is bad practice, because it massively bloats the github repo. You should always clear the outputs *before* doing a git commit.  The CI will execute the notebook so that the end result web site still has the outputs. Post facto this might be challenging to clear out all your old commits, though... do you have enough of a handle on git for how to do this, or should we think about alternatives?
* The use of a single-author package from a github repo as a dependency is strongly discouraged for a few reasons: 1) it basically means at any time this notebook might be broken by someone fiddling with the package on github.  I get that this is *your* repo, but the principle still applies since in the future people other than you might be working on this notebook.  2) the "meat" of your model is buried in the smartsnn package, so a reader has to go hunt down in an undocumented github repo what the model actually looks like. There are a few possible fixes for this, but the one I would recommend is simply: copy the implementation of the ConvNet and Laplacian_NLL classes over to this notebook (and if you like, put a comment at the top saying where you copied them from).  This is normally not good practice to copy-and-paste code instead of using a dependency, but in this case a large point of this notebook is pedagogical: that is, we *want* the user to see the structure of the model and the loss function so they can understand how it's done, if the code is reasonably straightforward (and it is - e.g. the training function in this notebook is bigger than the `ConvNet` class)

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+numpy
```
[the remark this answers is not in the document]
Comment (zclaytor) [reply notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/requirements.txt:1]: Done.

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+matplotlib
+torch
+astropy
+git+https://github.com/zclaytor/smartsnn
```
[the remark this answers is not in the document]
Comment (zclaytor) [reply notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/requirements.txt:5]: Since the `smartsnn` code is lightweight, and as you mentioned this way is more pedagogical, I've put the CNN code directly into the notebook. I like it much more this way.

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb [hunk 7]
```diff
@@ -0,0 +1,1019 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e3e8db64",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "# Install needed packages\n",
```
[the remark this answers is not in the document]
Comment (zclaytor) [reply notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:90]: Removing this cell - thanks!

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb [hunk 8]
```diff
@@ -0,0 +1,1019 @@
...
+    "\n",
+    "# For tensor and CNN operations\n",
+    "import torch \n",
+    "import torch.optim as optim\n",
+    "from torch.optim.lr_scheduler import ReduceLROnPlateau\n",
+    "from torch.utils.data import Dataset, DataLoader\n",
+    "\n",
+    "# For defining and training the CNN\n",
+    "from smartsnn.model import ConvNet\n",
+    "from smartsnn.model import Laplacian_NLL as loss_function"
```
[the remark this answers is not in the document]
Comment (zclaytor) [reply notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:116]: I've removed references to `smartsnn` since the code is now _in_ the notebook.

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb [hunk 9]
```diff
@@ -0,0 +1,1019 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "766eb96b",
+   "metadata": {},
+   "source": [
+    "## 1. Define training data loader\n",
+    "We use Pytorch `Dataset` (subclassed here) and `DataLoader` (used below) to load the wavelet data. Pytorch `Dataset` and `DataLoader` can be accessed in chunks, which makes training more efficient."
```
[the remark this answers is not in the document]
Comment (zclaytor) [reply notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:126]: Change incorporated.

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb [hunk 10]
```diff
@@ -0,0 +1,1019 @@
...
+    "            wmax = np.nan\n",
+    "            p = p.astype(np.float32)\n",
+    "            w = w.astype(np.uint8)\n",
+    "        self.pmax = pmax\n",
+    "        self.wmax = wmax\n",
+    "\n",
+    "        # determine how many examples to use for each partition\n",
+    "        n_train, n_val, n_test = (max_n * np.array(split)/np.sum(split)).astype(int).cumsum()\n",
+    "\n",
+    "        if mode == \"train\":\n",
```
[the remark this answers is not in the document]
Comment (zclaytor) [reply notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:190]: I played around with the random_split function, and while I agree it would require less code, it doesn't seem to let me (1) only use a subset of the training data (it yells at me if the requested lengths do not sum to the total number of rows in the input data) OR (2) create the dataloaders using the memorymap, so even loading 10,000 training examples requires first reading all 1 million training examples, which is unnecessarily memory intensive.

I'm absolutely open to better options if you know of any, though.

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb [hunk 11]
```diff
@@ -0,0 +1,1019 @@
...
+   "id": "034d9728",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train(model, train_loader, val_loader, num_epochs=100, early_stopping_patience=10, device=torch.device(\"cpu\")):\n",
+    "    \"\"\"Train the neural network for all desired epochs.\n",
+    "    \"\"\"\n",
+    "    optimizer = optim.Adam(model.parameters(), lr=1e-5)\n",
+    "    # Set learning rate scheduler\n",
+    "    scheduler = ReduceLROnPlateau(optimizer, factor=0.7, patience=3)\n",
```
[the remark this answers is not in the document]
Comment (zclaytor) [reply notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:256]: Agreed - explanation added.

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb [hunk 12]
```diff
@@ -0,0 +1,1019 @@
...
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "2094d10a",
+   "metadata": {},
+   "source": [
+    "## 5. Compare the predictions to the true values\n",
+    "and note that since we're only training over a small fraction of the training set in this tutorial, we don't necessarily expect the predictions to match the true values."
```
[the remark this answers is not in the document]
Comment (zclaytor) [reply notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:832]: I now more explicitly say "if you use a larger dataset (and train for longer) it will do better", and I point the user to the exact location and variables to change to do so.

File: notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb (hunk 3 above)
Comment (zclaytor) [reply notebooks/hello-universe/Regressing_TESS_rotation_periods_with_CNNs/Regressing_TESS_rotation_periods_with_CNNs.ipynb:955]: Agreed; I've reworded the discussion to present RMS error as an alternative, but still useful, accuracy metric.

**Comment by zclaytor** [comment]
I had to do a rebase and manually clear the notebook outputs commit by commit, which... whew... but I think it's done... but let me know if something still looks wrong.
